Client, Blockchain: Fix Goerli Clique Difficulty Bug

holgerd77 opened the pull request:
Fixes #1100

Got it, cause for the difficulty bug was the ordering of the signers mentioned in the issue. 😄 

Now there is an additional clique bug triggered some blocks down the road though, now with a pretty complex voting scenario. Here is the output:

![image](https://user-images.githubusercontent.com/931137/107795031-76a4e600-6d58-11eb-9621-833d972c2bc1.png)

holgerd77 commented:
(PR can also be merged if convenient, as you like)

jochem-brouwer commented:
There's a blockchain test failing? I suspect that in this case this is one of these cases where the test is actually wrong, and not the code?

holgerd77 commented:
Ah. No? Where are you reading this information from?

jochem-brouwer commented:
[This](https://github.com/ethereumjs/ethereumjs-monorepo/pull/1103/checks?check_run_id=1888723885) CI?

holgerd77 commented:
Ah, non, that are just some test failures I haven't fixed yet caused by the clique fixes on the first bug. That's easily fixable and totally unrelated to the second bug I have posted here. This is again about clique voting and signer updates, still something wrong in the implementation there.

ryanio commented on packages/blockchain/src/index.ts:
```diff
@@ -483,6 +487,12 @@ export default class Blockchain implements BlockchainInterface {
 
     await this.dbManager.batch(dbOps)
+    // Output active signers for debugging purposes
+    let i = 0
+    for (const signer of this.cliqueActiveSigners()) {
+      debug(`Clique signer [${i}]: ${signer}`)
+      i++
+    }
```
nice little es6 syntax destructuring trick for getting the index of an iteration in a for-of loop:
```suggestion
    for (const [i, signer] of this.cliqueActiveSigners().entries()) {
      debug(`Clique signer [${i}]: ${signer}`)
    }
```

jochem-brouwer replied:
Wow, cool 😄

holgerd77 commented:
Two things I noticed here:

1) The implementation of the vote count with [alreadyVoted](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/blockchain/src/index.ts#L526) doesn't seem correct to me. A vote from the past may not have been affected the majority consensus. If there has been a change on the votes (or the signer list) in between a double vote, this second vote may very well cause a new consensus situation, it just shouldn't been counted twice. So completely ignoring this second (or third, fourth,...) vote is likely not correct. Also from the spec:

> Only the latest proposal per target beneficiary is kept from a single signer.

This contradicts the current implementation.

2) I had the case when testing with Goerli that blockchain sync and VM execution were (as they often are) out of sync. When there is a change on the active signers in between, the block validation triggered in the VM will not work correctly for the part using the blockchain object (so e.g. the clique difficulty validation). I would suggest that we therefore skip the block validation in the VM since this is redundant anyhow with the blocks having already been validated on sync.

//cc @ryanio

ryanio commented:
I added some commits to resolve the failing blockchain tests and rebased on master.

---

> 1. The implementation of the vote count with [alreadyVoted](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/blockchain/src/index.ts#L526) doesn't seem correct to me. A vote from the past may not have been affected the majority consensus. If there has been a change on the votes (or the signer list) in between a double vote, this second vote may very well cause a new consensus situation, it just shouldn't been counted twice. So completely ignoring this second (or third, fourth,...) vote is likely not correct. Also from the spec:
> 
> > Only the latest proposal per target beneficiary is kept from a single signer.
> 
> This contradicts the current implementation.

Ok, I can work on this.

> 2. I had the case when testing with Goerli that blockchain sync and VM execution were (as they often are) out of sync. When there is a change on the active signers in between, the block validation triggered in the VM will not work correctly for the part using the blockchain object (so e.g. the clique difficulty validation). I would suggest that we therefore skip the block validation in the VM since this is redundant anyhow with the blocks having already been validated on sync.

Ah ok, should we configure the client's blockchain if (poa) then turn validateBlocks off?

holgerd77 commented:
> I added some commits to resolve the failing blockchain tests and rebased on master.

Thanks for fixing the tests. 😄 

> > This contradicts the current implementation.
> 
> Ok, I can work on this.

Also experimented a bit on this yesterday, came up with the following code:

```typescript
let beneficiaryVotes
// AUTH vote
if (nonce.equals(CLIQUE_NONCE_AUTH)) {
  beneficiaryVotes = this._cliqueLatestVotes.filter(
    (vote) =>
      vote[0].gte(lastEpochBlockNumber) &&
      vote[1][1].equals(beneficiary) &&
      vote[1][2].equals(CLIQUE_NONCE_AUTH)
  )
  // Majority consensus
  if(beneficiaryVotes.length - numAlreadyVoted >= limit) {
    // Authorize new signer
    activeSigners.push(beneficiary)
    activeSigners.sort((a, b) => {
      // Sort by buffer size
      return a.toBuffer().compare(b.toBuffer())
    })
    // Discard votes for added signer
    this._cliqueLatestVotes = this._cliqueLatestVotes.filter(
      (vote) => !vote[1][1].equals(beneficiary)
    )
  }
// DROP vote
} else {
  beneficiaryVotes = this._cliqueLatestVotes.filter(
    (vote) =>
      vote[0].gte(lastEpochBlockNumber) &&
      vote[1][1].equals(beneficiary) &&
      vote[1][2].equals(CLIQUE_NONCE_DROP)
  )
  // Majority consensus
  if(beneficiaryVotes.length - numAlreadyVoted >= limit) {
    // Drop signer
    activeSigners = activeSigners.filter((signer) => !signer.equals(beneficiary))
    // Discard votes from removed signer
    this._cliqueLatestVotes = this._cliqueLatestVotes.filter(
      (vote) => !vote[1][0].equals(beneficiary)
    )
  }
}
if (beneficiaryVotes.length - numAlreadyVoted >= limit) {
  debug(
    `[Block ${header.number.toNumber()}] Clique majority consensus -> update signer states`
  )
  const newSignerState: CliqueSignerState = [header.number, activeSigners]
  await this.cliqueUpdateSignerStates(newSignerState)
}
```

So replacing the respective code parts before. This *did* solve the first Goerli failure on block 5310 posted in the first screenshot. I went into some new issues some blocks further though (seems clique edge cases have been really triggered on purpose along these first blocks, which ultimately is a good thing 😄 ). Also some additional tests were failing. So I hesitated pushing directly. Might serve as some inspiration though.

Maybe I'll also find some time to continue to work on this during the day.

> Ah ok, should we configure the client's blockchain if (poa) then turn validateBlocks off?

I would think we can generally turn this off for the VM since locks are already validated in the blockchain class and this is just executing the same code twice. Or am I missing something?

holgerd77 commented on packages/blockchain/src/index.ts:
```diff
@@ -483,6 +487,10 @@ export default class Blockchain implements BlockchainInterface {
+    // Output active signers for debugging purposes
+    for (const [i, signer] of this.cliqueActiveSigners().entries()) {
+      debug(`Clique signer [${i}]: ${signer}`)
+    }
```
Yeah, this is really nice stuff. 😄

holgerd77 commented on packages/blockchain/test/clique.spec.ts:
```diff
@@ -44,61 +44,61 @@ tape('Clique: Initialization', (t) => {
 
   const C: Signer = {
-    address: new Address(Buffer.from('8458f408106c4875c96679f3f556a511beabe138', 'hex')),
+    address: new Address(Buffer.from('83c30730d1972baa09765a1ac72a43db27fedce5', 'hex')),
```
Lol, I also started fixing the tests yesterday but I went along with changing the order of the signers in the test cases. This is obviously the significantly better idea. 😎

holgerd77 commented:
@ryanio ok, I've reworked this significantly, I still not have got 1-2 corner cases under control unfortunately. I nevertheless think that the reworked structure provides a better ground for further refinements, I therefore dared to push.

One test is failing (again), nevertheless this better takes this touched-account scenario in consideration, should be a relatively simple fix, then things should be working (hopefully).

Feel free to pick up.

holgerd77 commented:
(phew. but hard stuff. 😕 )

holgerd77 commented:
(maybe keep a local copy of your branch just in case, I am getting so confused by this stuff here)

holgerd77 commented:
On this latest state the tests pass again, but Goerli run fails at:

![image](https://user-images.githubusercontent.com/931137/107851944-c5f91e00-6e0d-11eb-9887-89aa9cdd8fc4.png)

holgerd77 commented:
Yeah, yeah, yeah, got it. 🎉

![image](https://user-images.githubusercontent.com/931137/107852920-9e0cb900-6e13-11eb-80d0-a6c2313c1200.png)

Really spent another 3-4 hours debugging on this, would have never imagined.

What a journey, phew

ryanio reviewed (approved):
nice work, great, lgtm!

holgerd77 commented on packages/blockchain/src/index.ts:
```diff
@@ -498,71 +508,115 @@ export default class Blockchain implements BlockchainInterface {
+            beneficiaryVotesDROP.push(vote[1][0])
+          }
+        }
+        let numBeneficiaryVotesDROP = beneficiaryVotesDROP.length
```
@ryanio thanks for review, wasn't so sure about here how to do this elegantly, with a direct additional `filter()` or `reduce()` call, if you know a better way feel free to rewrite on some subsequent PR (then we really make sure that it still syncs these first 6.000 or so Goerli blocks though, a bit trick. I was wondering if we should extract the case which was failing for us before the rewrite and add (propose) as an additional test case to the EIP. I am planning to do a PR with some suggestions for improvements and clarifications anyhow once we are stable (might be more or less: now, we'll see, likely though), have taken notes along the lines. If you have something to add feel free to send me. We can alternatively also fork the EIP repo to `ethereumjs` and do the PR together).
